opengrok-web/src/test/java/org/opengrok/web/api/v1/controller/ConcurrentConfigurationControllerTest.java

```diff
@@ -0,0 +1,188 @@
+
+    private static final int PROJECTS_COUNT = 20;
+    private static final int THREAD_COUNT = Math.max(30, Runtime.getRuntime().availableProcessors() * 2);
+    private static final int TASK_COUNT = 100;
+
+    @Rule
+    public ConditionalRunRule rule = new ConditionalRunRule();
+
+    @Mock
+    private SuggesterService suggesterService;
+
+    private RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+    TestRepository repository;
+    String origSourceRootPath;
+    String origDataRootPath;
+    Map<String, Project> origProjects;
+    List<RepositoryInfo> origRepositories;
+
+    @Override
+    protected Application configure() {
+        MockitoAnnotations.initMocks(this);
+        return new ResourceConfig(ConfigurationController.class)
+                .register(new AbstractBinder() {
+                    @Override
+                    protected void configure() {
+                        bind(suggesterService).to(SuggesterService.class);
+                    }
+                });
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        origSourceRootPath = env.getSourceRootPath();
+        origDataRootPath = env.getDataRootPath();
+        origProjects = env.getProjects();
+        origRepositories = env.getRepositories();
+
+        // prepare test repository
+        repository = new TestRepository();
+        repository.create(HistoryGuru.class.getResourceAsStream("repositories.zip"));
+
+        env.setSourceRoot(repository.getSourceRoot());
+        env.setDataRoot(repository.getDataRoot());
+
+
+        List<RepositoryInfo> repositoryInfos = new ArrayList<>();
+        Map<String, Project> projects = new TreeMap<>();
+
+        /*
+         * Prepare PROJECTS_COUNT git repositories, named project-{i} in the test repositories directory.
+         */
+        for (int i = 0; i < PROJECTS_COUNT; i++) {
+            Project project = new Project();
+            project.setName("project-" + i);
+            project.setPath("/project-" + i);
+            RepositoryInfo repo = new RepositoryInfo();
+            repo.setDirectoryNameRelative("/project-" + i);
+
+            projects.put("project-" + i, project);
+            repositoryInfos.add(repo);
+
+            // create the repository
+            FileUtils.copyDirectory(
+                    Paths.get(repository.getSourceRoot(), "git").toFile(),
+                    Paths.get(repository.getSourceRoot(), "project-" + i).toFile()
+            );
+        }
+
+        env.setRepositories(repositoryInfos);
+        env.setProjects(projects);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+        repository.destroy();
+
+        env.setProjects(origProjects);
+        env.setRepositories(origRepositories);
+        env.setSourceRoot(origSourceRootPath);
+        env.setDataRoot(origDataRootPath);
+    }
+
+    @Test
+    @ConditionalRun(RepositoryInstalled.GitInstalled.class)
```

Comment:
can you annotate the whole class as being Git dependent ? The `setUp()` copies the repositories (rather than cloning them which would be arguably better) however this would be of no use if the tests cannot be run.

Reply:
ok